src/internal/TokenRegistry.ts

```diff
@@ -0,0 +1,78 @@
+import { Tokens } from ".";
+
+type TokenMetadata = {
+  address: string;
+  symbol: string;
+  decimals: number;
+  chainId: number;
+};
+
+type TokenDetails = {
+  address: string;
+  decimals: number;
+};
+
+export default class TokenRegistry {
+  private static instance: TokenRegistry;
+
+  private addressMap: Map<string, TokenMetadata>;
+  private symbolMap: Map<string, TokenMetadata>;
+
+  constructor(tokens: TokenMetadata[]) {
+    this.addressMap = new Map();
+    this.symbolMap = new Map();
+
+    for (const token of tokens) {
+      const normalizedAddress = this.createKey(token.address, token.chainId);
+      const normalizedSymbol = this.createKey(token.symbol, token.chainId);
+
+      this.addressMap.set(normalizedAddress, token);
+      this.symbolMap.set(normalizedSymbol, token);
+    }
+  }
+
+  static getInstance(): TokenRegistry {
+    if (!TokenRegistry.instance) {
+      const tokens: TokenMetadata[] = Tokens.tokens
+      TokenRegistry.instance = new TokenRegistry(tokens);
+    }
+    return TokenRegistry.instance;
+  }
+
+  private normalize(value: string): string {
+    return value.toLowerCase().trim();
+  }
+
+  private createKey(value: string, chainId: number): string {
+    return `${this.normalize(value)}-${chainId}`;
+  }
+
+  getByAddressOrSymbol(addressOrSymBol: string, chainId: number): TokenDetails {
```

Comment:
typo `addressOrSymBol` should be `addressOrSymbol`

Reply:
generally our codebase uses single object named args for functions. To remain consistent, this would be: `{ addressOrSymbol: string, chainId: number }`. I'd recommend updating all functions to this, but most importantly the public ones